### ISSUE TITLE
failure if upgraded vatParameters includes an imported vpid that was previously watched

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#9728]
## Describe the bug

While investigating an orch/async-flow crash, @mhofman and I identified a potential problem in the way watched promises are handled. We haven't seen exactly this crash, but the symptom would be an error during vat upgrade, `revivePromise called on pre-existing ${slot}`, with a stack trace that includes `revivePromise` and `loadWatchedPromiseTable` and `startVat`.

To trigger this, the necessary conditions are:

* the first incarnation imports and watches a promise, 
  * we'll call this promise `p-10`, after a likely vpid
* the first incarnation also calls `VatData.watchPromise(p, watcher)`
  * The watch doesn't happen in `buildRootObject` or it would trigger #9732, which is a variation of this issue
* the parent vat then does `E(adminNode).upgradeVat(bundleCap, { vatParameters: p })`, or any other vatParameters that contains `p` somehow

What happens next:
* the kernel does `processUpgradeVat`, which shuts down the old incarnation, rejects (disconnects) all its decided promises, then delivers a `startVat` to the new incarnation
* inside the vat, at the beginning of startVat, liveslots will deserialize `vatParametersCapData` to produce the `vatParameters` it will pass into `buildRootObject`
  * this adds the `p-10` VPID to the `slotToVal`/`valToSlot` table
  * it also gets registered in `importedVPIDs`, so a subsequent `dispatch.notify` will resolve it
  * it also performs a `syscall.subscribe('p-10')`, which is redundant but idempotent
* then, buildRootObject is called, and userspace re-defines all its Kinds, including the handler used by the earlier `watchPromise`
  * `buildRootObject` does not call `watchPromise` on `p` or that would create a registration and trigger #9732, a slightly different version of this issue.
* then, after buildRootObject is done, liveslots calls `watchedPromiseManager.loadWatchedPromiseTable()` to re-establish the watches defined by the first incarnation
  * this walks the list of registered VPIDs, and calls `revivePromise(vpid)` on each one, to get a real Promise object, to which it can use `E.when()` to attach the code that calls the durable handlers
  * `revivePromise()` throws an error when it sees that the VPID is already associated with a Promise:

https://github.com/Agoric/agoric-sdk/blob/81ac38153db929618831ac19f57c019769d3cbef/packages/swingset-liveslots/src/liveslots.js#L793

In a perfect world, I'd impose a restriction that `vatParameters` can only be durable data (thus excluding Promises), but @mhofman pointed out that this ship sailed a long time ago, and we routinely pass Promises into contracts via `vatParameters.privateArgs`.

Note that this problem can also happen for promises that were created by the previous incarnation. These promises cannot currently be included in vatParameters (or any message argument) because of #9747, but once that issue is fixed, the previously exported and watched promises would be equivalent to an imported watched promise.

The fix will be to relax `revivePromise` and make it tolerant of already-deserialized Promises. Something like:

```diff
diff --git a/packages/swingset-liveslots/src/liveslots.js b/packages/swingset-liveslots/src/liveslots.js
index 7246f41f2..1c7012953 100644
--- a/packages/swingset-liveslots/src/liveslots.js
+++ b/packages/swingset-liveslots/src/liveslots.js
@@ -790,7 +790,19 @@ function build(
     meterControl.assertNotMetered();
     const { type } = parseVatSlot(slot);
     type === 'promise' || Fail`revivePromise called on non-promise ${slot}`;
-    !getValForSlot(slot) || Fail`revivePromise called on pre-existing ${slot}`;
+    const val = getValForSlot(slot);
+    if (val) {
+      return val;
+    }
     const pRec = makePipelinablePromise(slot);
     importedVPIDs.set(slot, pRec);
     const p = pRec.promise;
```

We also need a specific test for this.

**Comment by mhofman** [comment]
Assigned myself since I have been working on a PR with tests for this.

**Comment by mhofman** [comment]
Updated the issue as this can happen for imported promises as well, which is a case less convoluted than previously exported promises, and more importantly it wouldn't trip #9747 first.

**Comment by warner** [comment]
note: while looking at #10270, we realized that promises are *not* allowed in `vatParameters`, because they must pass through a device (`device-vat-admin`), and devices don't do promises, and the liveslots `D()` call has explicit checks to prevent `p`-type vrefs from being passed either into the device (as arguments) or back out again (as a result).

The kernel may or may not handle promises in `vatParameters` gracefully, we don't know. (I can't think of any reason why not: they get a refcount, liveslots will subscribe to them during `startVat` when it demarshals the capdata, that might cause a duplicate subscription but I *think* the kernel is just updating a set, and if the promise was resolved while the incarnation wasn't running, the `subscribe` should trigger a `notify`).
